Lab4 Numeronym function

brnbke opened the pull request:
Fixes #402

Review of colleague's PR #375 

Creating Numeronym function takes a slice of stings, return a slice of strings that's a
Numeronym of each input for string > 2 runes in length

Commit:
Lab4 Numeronym function

Takes a slice of stings, return a slice of strings that's a
Numeronym of each input for string > 2 runs in length

runnerdave reviewed (approved):
looks good to me, nice comprehensive and creative tests!

camscale commented on 04_numeronym/brnbke/main.go:
```diff
@@ -0,0 +1,29 @@
+		l := utf8.RuneCountInString(val)
+		if l > 3 {
+			s := []rune(val)[0]   // starting character
+			e := []rune(val)[l-1] // ending character
```
a suggestion not required to implement, since you are already importing `unicode/utf8`: use `utf8.DecodeRuneInString()` and `utf8.DecodeLastRuneInString()`: https://golang.org/pkg/unicode/utf8/

brnbke replied:
Implemented, it's considerably more efficient to use this also if my speed test was right

camscale commented on 04_numeronym/brnbke/main.go:
```diff
@@ -0,0 +1,29 @@
+		if l > 3 {
+			s := []rune(val)[0]   // starting character
+			e := []rune(val)[l-1] // ending character
+			result = append(result, fmt.Sprintf("%c%d%c", s, l-2, e))
```
suggestion: do the `append` once outside the `else` below (get rid of the else) by changing this to:
```
val = fmt.Sprintf("%c%d%c", s, l-2, e)
```
since conceptually what you're doing in this branch is changing the value. An `append` is done for all values.

brnbke replied:
Fair point, actually makes it easier to read also, implemented.

camscale commented on 04_numeronym/brnbke/main_test.go:
```diff
@@ -0,0 +1,63 @@
+				"🏦🔫🗯💰🚗😬🚓🚓🚓🚓😱🔫🔫🔫🚓🚓🚒🚑😨😢😰😢😢🚓🏛😢😢😢😢⏸🔒👮",
+				"ηλεκτροεγκεφαλογράφημα",
+				"فسيكفيكهم"},
+			want: []string{"⽉7⽖", "🏦30👮", "η20α", "ف7م"}},
```
food for thought: should the first result actually be ⽉七⽖ (I've assumed Japanese but I really only know English and computer languages).

brnbke replied:
You might be right, to be honest I just stole the text from the previous ex and didn't really think about i18n for the numbers...

camscale reviewed (dismissed):
All good. Just some suggestions - up to you if you want to do them.

brnbke committed:
PR changes - using DecodeRuneInString & DecodeLastRuneInString

getting the first and last run in a more efficient manner than
using []rune(val)[0] and cleaning up append

camscale reviewed (approved):
Two comments, but go ahead and merge anyway:
* Your PR says length > 2 but the code says length > 3
* You should have a test with string of length 4 - that's the boundary condition. You should verify that strings of length 3 are not perverted but strings of length 4 are.

But the Go code is all good, so merge away. Please squash merge so there are no fixup commits merged to master.

brnbke committed:
Merge branch 'master' into lab4
